sweep/sweeper.go

```diff
@@ -441,8 +574,65 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch,
 
 			bestHeight = epoch.Height
 
-			log.Debugf("New blocks: height=%v, sha=%v",
-				epoch.Height, epoch.Hash)
+			log.Debugf("New block: height=%v, sha=%v", epoch.Height,
+				epoch.Hash)
+
+			// In order to correctly meet the confirmation targets
+			// of inputs, we'll bump their fee rates accordingly.
+			//
+			// NOTE: Meeting the confirmation target is a bit
+			// inaccurate at the moment due to the sweeper randomly
+			// rescheduling publish attempts for inputs.
+			for input, confTarget := range s.confTargetInputs {
+				pendInput := s.pendingInputs[input]
+
+				// We'll only proceed to query for a new fee
+				// rate if we don't exceed the threshold of
+				// confirmations left in order to do so.
+				expectedConfHeight := pendInput.initialHeight +
+					confTarget
+				numBlocksLeft := expectedConfHeight - bestHeight
+				if numBlocksLeft < 1 {
+					numBlocksLeft = 1
+				}
+				threshold := int32(
+					float64(confTarget) * confTargetRatio,
+				)
+
+				if numBlocksLeft > threshold {
+					continue
+				}
+
+				newFeeRate, err := s.cfg.FeeEstimator.EstimateFeePerKW(
+					uint32(numBlocksLeft),
+				)
+				if err != nil {
+					log.Errorf("Unable to determine new "+
+						"fee rate for input %v: %v",
+						pendInput.input.OutPoint(),
+						err)
+					continue
+				}
+
+				// We're not able to lower the fee rate of
+				// inputs, so we'll skip any lower ones.
+				if newFeeRate <= pendInput.feeRate {
+					log.Debugf("Ignoring lower fee rate "+
+						"of %v for input %v with fee "+
+						"rate %v", newFeeRate,
+						pendInput.input.OutPoint(),
+						pendInput.feeRate)
+					continue
+				}
+
+				log.Debugf("Applying new fee rate of %v for "+
+					"input %v in order to confirm by "+
+					"height %v", newFeeRate,
+					pendInput.input.OutPoint(),
+					expectedConfHeight)
+
+				pendInput.feeRate = newFeeRate
+			}
```

Comment:
With the `getFeeRate()` suggestion, this would all be moved behind that call back and abstracted away from `UtxoSweeper`.

Reply:
Could also be a method on the `FeePreference` struct.